File: src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionMaxTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import com.google.gson.JsonNull;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class TransformerFunctionMaxTest extends BaseTest {
    @Test
    void inline() {
        var arr = new Object[] {4,-2,13.45, JsonNull.INSTANCE};
        assertTransformation(arr, "$$max($$long:40):$", fromJson("40"));
        assertTransformation(arr, "$$max(-8,NUMBER):$", fromJson("13.45"));
        assertTransformation(arr, "$$max():$", fromJson("13.45"));
        assertTransformation(arr, "$$max(z,STRING):$", fromJson("z"));

    }

    record holder(BigDecimal value) {}
    @Test
    void object() {
        var arr = new holder[] {new holder(new BigDecimal(4)),new holder( new BigDecimal(2)),new holder( new BigDecimal("13.45")), new holder(null)};
        assertTransformation(arr, fromJson("""
{
  "$$max":"$","by":"##current.value","default":"zz","type":"STRING"
}
"""), fromJson("'zz'"));

        assertTransformation(arr, fromJson("""
{
  "$$max":"$","by":"##current.value"
}
"""), fromJson("13.45"));



    }

}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionMax.test.ts
import type BigNumber from "bignumber.js";
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";
import { BigDecimal } from "../../functions/common/FunctionHelpers";

describe("TransformerFunctionMax", () => {
  test("inline", async () => {
    const arr = [4, -2, 13.45, null];
    await assertTransformation(arr, "$$max($$long:40):$", BigInt(40));
    await assertTransformation(arr, "$$max(-8,NUMBER):$", 13.45);
    await assertTransformation(arr, "$$max():$", 13.45);
    await assertTransformation(arr, "$$max(z,STRING):$", "z");
  });

  class Holder {
    public value: BigNumber | null;
    constructor(value: BigNumber | null) {
      this.value = value;
    }
  }

  test("object", async () => {
    const arr = [
      new Holder(BigDecimal(4)),
      new Holder(BigDecimal(2)),
      new Holder(BigDecimal("13.45")),
      new Holder(null),
    ];
    await assertTransformation(
      arr,
      {
        $$max: "$",
        by: "##current.value",
        default: "zz",
        type: "STRING",
      },
      "zz",
    );

    await assertTransformation(
      arr,
      {
        $$max: "$",
        by: "##current.value",
      },
      BigDecimal("13.45"),
    );
  });
});
